protocols/relay/src/v2/client.rs

```diff
@@ -0,0 +1,561 @@
+                                    event: handler::In::Reserve { to_listener },
+                                });
+                        }
+                        None => {
+                            self.rqsts_pending_connection
+                                .entry(relay_peer_id)
+                                .or_default()
+                                .push(RqstPendingConnection::Reservation { to_listener });
+                            let handler = self.new_handler();
+                            return Poll::Ready(NetworkBehaviourAction::Dial {
+                                opts: DialOpts::peer_id(relay_peer_id)
+                                    .addresses(vec![relay_addr])
+                                    .extend_addresses_through_behaviour()
+                                    .build(),
+                                handler,
+                            });
+                        }
+                    }
+                }
+                Poll::Ready(Some(transport::TransportToBehaviourMsg::DialReq {
+                    relay_addr,
+                    relay_peer_id,
+                    dst_peer_id,
+                    send_back,
+                    ..
+                })) => {
+                    match self
+                        .connected_peers
+                        .get(&relay_peer_id)
+                        .and_then(|cs| cs.get(0))
+                    {
+                        Some(connection_id) => {
+                            return Poll::Ready(NetworkBehaviourAction::NotifyHandler {
+                                peer_id: relay_peer_id,
+                                handler: NotifyHandler::One(*connection_id),
+                                event: handler::In::EstablishCircuit {
+                                    send_back,
+                                    dst_peer_id,
+                                },
+                            });
+                        }
+                        None => {
+                            self.rqsts_pending_connection
+                                .entry(relay_peer_id)
+                                .or_default()
+                                .push(RqstPendingConnection::Circuit {
+                                    dst_peer_id,
+                                    send_back,
+                                });
+                            let handler = self.new_handler();
+                            return Poll::Ready(NetworkBehaviourAction::Dial {
+                                opts: DialOpts::peer_id(relay_peer_id)
+                                    .addresses(vec![relay_addr])
+                                    .extend_addresses_through_behaviour()
+                                    .build(),
+                                handler,
+                            });
+                        }
+                    }
+                }
+                Poll::Ready(None) => unreachable!(
```

Comment:
Why is this unreachable? Is it because the transport is gauranteed to outlive (or have the same lifetime as) the behavior?

Reply:
Yes. The `ClientTransport` is guaranteed to live as long as the `Client` `NetworkBehaviour`. I.e. the `Client` should never be polled via `NetworkBehaviour::poll` after the `ClientTransport` has been dropped.

What wording would make this more clear for a newcomer?

```suggestion
                Poll::Ready(None) => unreachable!(
                    "`Relay` `NetworkBehaviour` polled after channel from \
                     `RelayTransport` has been closed. Unreachable under \
                     the assumption that the `Client` is never polled after \
                     `ClientTransport` is dropped."
```

Reply:
👍 Excellent!

Reply:
Done via c0c088055fadc2a69ffc4923c608e17f864a997e.